### ISSUE TITLE
no warning for div by zero:    let x = 1 / false as u8;

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#49760]
inlined inside a ````println!()````, I get a warning:
````rust
fn main() {
    println!("{}", 1/false as u8);
}
````
````
error[E0080]: constant evaluation error
 --> src/main.rs:4:17
  |
4 |     println!("{}", 1/false as u8);
  |                    ^^^^^^^^^^^^^ attempted to do overflowing math
error: aborting due to previous error
````
however, when I involve a variable
````rust
fn main() {
	let x = 1 / false as u8;
	println!("{}", x);
}
````
there is no warning at all but looking at the generated code it is quite clear that this panics:
https://godbolt.org/g/NQmSFM

**Comment by rkruppe** [comment]
It warns on current stable, but not on beta and nightly. Might be miri fallout? cc @oli-obk 

Curiously, when you take a reference (to promote rvalue promotion to static) you get a hard error on beta and nightly (stable is still a warning). Isn't that a regression?

**Comment by nikomatsakis** [comment]
triage: P-medium

Still want to get those warnings, but no longer P-high

**Comment by matthiaskrgr** [comment]
Hm, what  is the status of this?

The PR was merged, but with ````rustc 1.27.0-nightly (e5f80f2a4 2018-05-09)````, I don't get a compiler warning for the first example.

For the second one I get two though? :confused: 
````
   Compiling zero1 v0.1.0 (file:///tmp/zero1)
warning: constant evaluation error
 --> src/main.rs:3:17
  |
3 |     println!("{}", 1 / false as u8);
  |                    ^^^^^^^^^^^^^^^ attempt to divide by zero
  |
  = note: #[warn(const_err)] on by default
warning: constant evaluation error
 --> src/main.rs:3:17
  |
3 |     println!("{}", 1 / false as u8);
  |                    ^^^^^^^^^^^^^^^ attempt to divide by zero
    Finished dev [unoptimized + debuginfo] target(s) in 0.48s
````

**Comment by oli-obk** [comment]
The PR was just for the compilation error. The missing warnings just need an impl for this branch:

https://github.com/rust-lang/rust/blob/master/src/librustc_mir/transform/const_prop.rs#L236

The warning duplication is unfortunate, but if we deduplicate it, we loose warnings in other edge cases. We could deduplicate it similar to the way the notes about where the lint level was set are deduplicated (note how the second warning has no `note: #[warn(const_err)] on by default` message). That would probably be the cleanest solution.
